tests/utils/log_reader.py

```diff
@@ -130,56 +157,67 @@ def _parse_line(self, line):
 
         return metric_name, metric_value
 
-    def get_metrics(self, names, timeout=10):
-        def get():
-            _metrics = dict()
-            for name in names:
-                metric_value = self._current_metrics.get(name)
-                if metric_value is None:
-                    return None
-                _metrics[name] = metric_value
-            return _metrics
-
-        start_time = time.time()
-        timeout_time = start_time + timeout
+    def _handle_new_line(self, line):
+        super(LogMetricReader, self)._handle_new_line(line)
+        name, value = self._parse_line(line)
+        self.current_metrics[name] = value
 
-        while not self._stop_event.is_set():
-            with self._lines_lock:
-                metrics = get()
-            if metrics is not None:
-                return metrics
+    def wait_for_metrics_exist(self, names, timeout=10):
+        timeout_time = time.time() + timeout
+        while True:
+            needed_metrics = dict()
+            for name in names:
+                value = self.current_metrics.get(name)
+                if value is None:
+                    break
+                needed_metrics[name] = value
+            else:
+                return needed_metrics
 
-            if time.time() >= timeout_time:
-                raise ValueError(
-                    "timeout of %s seconds reached while waiting for metrics"
-                    % (timeout)
-                )
+            try:
+                self.wait_for_next_line()
+            except LogReaderTimeoutError():
```

Comment:
Perhaps also log a message under debug or similar here - something along the line of "Timeout of x seconds reached when waiting for a new line, will retry"

Reply:
I've added line reading based on generators, so I could remove all those timeout checks.